evcache-core/src/main/java/net/spy/memcached/EVCacheMemcachedClient.java

```diff
@@ -124,62 +127,85 @@ private boolean isWrongKeyReturned(String original_key, String returned_key) {
     }
 
     public <T> EVCacheOperationFuture<T> asyncGet(final String key, final Transcoder<T> tc, EVCacheGetOperationListener<T> listener) {
-        final CountDownLatch latch = new CountDownLatch(1);
-        final EVCacheOperationFuture<T> rv = new EVCacheOperationFuture<T>(key, latch, new AtomicReference<T>(null), readTimeout.get().intValue(), executorService, client);
-        final Operation op = opFact.get(key, new GetOperation.Callback() {
-            private Future<T> val = null;
+        // we should only complete the latch when decode AND complete have completed
+        final CountDownLatch latch = new CountDownLatch(2);
 
-            public void receivedStatus(OperationStatus status) {
-                if (log.isDebugEnabled()) log.debug("Getting Key : " + key + "; Status : " + status.getStatusCode().name() + (log.isTraceEnabled() ?  " Node : " + getEVCacheNode(key) : "")
-                        + "; Message : " + status.getMessage() + "; Elapsed Time - " + (System.currentTimeMillis() - rv.getStartTime()));
-                try {
-                    if (val != null) {
-                        if (log.isTraceEnabled() && client.getPool().getEVCacheClientPoolManager().shouldLog(appName)) log.trace("Key : " + key + "; val : " + val.get());
-                        rv.set(val.get(), status);
-                    } else {
-                        if (log.isTraceEnabled() && client.getPool().getEVCacheClientPoolManager().shouldLog(appName)) log.trace("Key : " + key + "; val is null");
-                        rv.set(null, status);
-                    }
-                } catch (Exception e) {
-                    log.error(e.getMessage(), e);
-                    rv.set(null, status);
-                }
-            }
+        final EVCacheOperationFuture<T> rv = new EVCacheOperationFuture<>(key, latch, new AtomicReference<T>(null), readTimeout.get(), executorService, client);
+        final DistributionSummary dataSizeDS = getDataSizeDistributionSummary(
+                EVCacheMetricsFactory.GET_OPERATION,
+                EVCacheMetricsFactory.READ,
+                EVCacheMetricsFactory.IPC_SIZE_INBOUND);
+
+        final Operation op = opFact.get(key, new GetOperation.Callback() {
+            private volatile boolean decodingData = false;
+            private volatile OperationStatus status = null;
+            private final AtomicReference<T> valueHolder = new AtomicReference<>(null);
 
             @SuppressWarnings("unchecked")
             public void gotData(String k, int flags, byte[] data) {
+                decodingData = true;
+                if (isWrongKeyReturned(key, k)) {
+                    latch.countDown();
+                    return;
+                }
 
-                if (isWrongKeyReturned(key, k)) return;
+                boolean shouldLog = log.isDebugEnabled() && client.getPool().getEVCacheClientPoolManager().shouldLog(appName);
+                boolean alwaysSync = false;
```

Comment:
should be exposed via a fp to go via async path?

Reply:
fixed in 2b60f17